enhancements/proxy-cache.md

```diff
@@ -0,0 +1,124 @@
+* A Quay user can flush the cache to eliminate excess storage consumption.
+* Robot accounts can be created in the cache organization as usual, their RBAC can be managed for all existing cached repositories at a given point in time.
+* Provide metrics to track cache activity and efficiency (hit rate, size, evictions).
+
+### Non-Goals
+
+* In the first phase, configuring a cache proxy organization, caching upstreaming images, and quota management on cached repositories is the target.
+  Other goals will be implemented subsequently based on the work of this proposal.
+* Cached images are read-only.
+
+## Design Details
+
+The expected pull overview is depicted as below:
+![](https://user-images.githubusercontent.com/11522230/145866763-58f44c94-839b-4edb-a95b-b9c3648cf187.png)
+Design credits: @fmissi
+
+A user initiates a pulls of an image(say a `postgres:14` image) from an upstream repository on Quay. The repository is checked to see if the image is present.
+1. If the image does not exist, a fresh pull is initiated.
+  * The user is authenticated into the upstream registry and all the layers of `postgres:14` are pulled.
+  * The pulled layers are saved to cache and served to the user in parallel.
+  * This is depicted as below:
+    ![](https://user-images.githubusercontent.com/11522230/145871778-da01585a-7b1b-4c98-903f-809c214578da.png)
+    Design credits: @fmissi
+
+2. If the image exists in cache:
+  * A user can rely on Quay's cache to stay coherent with the upstream source so that I transparently get newer images from the cache
+     when tags have been overwritten in the upstream registry immediately or after a certain period of time.
+  * If the upstream image and cached version of the image are same:
+    * No layers are pulled from the upstream repository and the cached image is served to the user.
+
+  * If the upstream image and cached version of the image are different:
+    * The user is authenticated into the upstream registry and only the changed layers of `postgres:14` are pulled.
+    * The new layers are updated in cache and served to the user in parallel.
+  * This is depicted as below:
+    ![](https://user-images.githubusercontent.com/11522230/145872216-31350e08-6746-4e34-aebf-e59a7bf6b372.png)
+    Design credits: @fmissi
+
+3. If user initiates a pull when the upstream registry which is down:
+  * If the pull happens with the configured staleness period, the image stored in cache is served.
+  * If the pull happens after the configured staleness period, the error is propagated to the user.
+  * This is depicted as below:
+    ![](https://user-images.githubusercontent.com/11522230/145878373-c23d094b-709d-4859-b875-013ea33e34f7.png)
+    Design credits: @fmissi
+
+A quay admin can leverage quota of an organization to limit cache size so the backend storage consumption remains predictable
+by discarding images from the cache according to least recently used frequency.
```

Comment:
If the cache size is set to a very small number (1 MB), will Quay work as a proxy without caching or will it fail?

Reply:
Oh, I see, it'll overflow beyond quota limit.